### PR TITLE
Df pct change

### PR DESCRIPTION
**Pull request by Rubtsowa**

### CODE REVIEW

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 1]
```diff
@@ -87,7 +110,10 @@ def sdc_pandas_dataframe_reduce_columns(df, func_name, params, ser_params):
 
     df_func_name = f'_df_{func_name}_impl'
 
-    func_text, global_vars = _dataframe_reduce_columns_codegen(func_name, all_params, s_par, df.columns)
+    if func_name in funcs_result_series:
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: Extremely bad idea. Function behavior depends on some global object. Very opaque and confusing.

If you don't want to duplicate code it is better to add additional parameter to `sdc_pandas_dataframe_reduce_columns` function

File: examples/dataframe_pct_change.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+def dataframe_pct_change():
+    df = pd.DataFrame({"A": [14, 4, 5, 4, 1, 55],
+                       "B": [5, 2, 54, 3, 2, 32],
+                       "C": [20, 20, 7, 21, 8, 5],
+                       "D": [14, 3, 6, 2, 6, 4]})
+    out_df = df.pct_change()
+
+    return out_df
+
+# result DataFrame
```
Comment (AlexanderKalistratov) [inline examples/dataframe_pct_change.py]: Do we really need this comment? @PokhodenkoSA what do you think?

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 3]
```diff
@@ -96,6 +97,49 @@ def sdc_pandas_dataframe_reduce_columns(df, func_name, params, ser_params):
...
+        func_lines += [f'  series_{c} = pandas.Series(get_dataframe_data({func_params[0]}, {i}))',
+                       f'  {result_c} = series_{c}.{func_name}({series_params})']
+        result_name.append((f' {result_c}', c))
+
+    data = ', '.join(f'"{column_name}": {column}' for column, column_name in result_name)
+
+    func_lines += [f'  return pandas.DataFrame({{{data}}})']
+    func_text = '\n'.join(func_lines)
+
+    global_vars = {'pandas': pandas, 'np': numpy,
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: Looks like `numpy` isn't used at all. So please remove that.

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 4]
```diff
@@ -96,6 +97,49 @@ def sdc_pandas_dataframe_reduce_columns(df, func_name, params, ser_params):
...
+def sdc_pandas_dataframe_reduce_columns_df(df, func_name, params, ser_params):
+    all_params = ['df']
+    ser_par = []
+
+    for key, value in params.items():
+        all_params.append('{}={}'.format(key, value))
+    for key, value in ser_params.items():
+        ser_par.append('{}={}'.format(key, value))
+
+    s_par = '{}'.format(', '.join(ser_par[:]))
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: Why do you need:
1. `format`
2. copy of `ser_par` got via slice

I would say it's not required and have to be fixed.

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 5]
```diff
@@ -454,7 +498,7 @@ def count_overload(df, axis=0, level=None, numeric_only=False):
     Returns
     -------
     :obj:`pandas.Series` or `pandas.DataFrame`
-    for each column/row the number of non-NA/null entries. If level is specified returns a DataFrame.
+            for each column/row the number of non-NA/null entries. If level is specified returns a DataFrame.
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: I would decrease indentation to 4 spaces.

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 6]
```diff
@@ -96,6 +97,49 @@ def sdc_pandas_dataframe_reduce_columns(df, func_name, params, ser_params):
...
+    func_lines += [f'  return pandas.DataFrame({{{data}}})']
+    func_text = '\n'.join(func_lines)
+
+    global_vars = {'pandas': pandas, 'np': numpy,
+                   'get_dataframe_data': get_dataframe_data}
+
+    return func_text, global_vars
+
+
+def sdc_pandas_dataframe_reduce_columns_df(df, func_name, params, ser_params):
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: It's more like `apply` than `reduce`.

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 7]
```diff
@@ -96,6 +97,49 @@ def sdc_pandas_dataframe_reduce_columns(df, func_name, params, ser_params):
...
 
+def _dataframe_reduce_columns_codegen_df(func_name, func_params, series_params, columns):
+    result_name = []
+    joined = ', '.join(func_params)
+    func_lines = [f'def _df_{func_name}_impl({joined}):']
+    for i, c in enumerate(columns):
+        result_c = f'result_{c}'
+        func_lines += [f'  series_{c} = pandas.Series(get_dataframe_data({func_params[0]}, {i}))',
+                       f'  {result_c} = series_{c}.{func_name}({series_params})']
+        result_name.append((f' {result_c}', c))
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: Why do you need f-string here?

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 8]
```diff
@@ -127,11 +127,11 @@ def sdc_pandas_dataframe_reduce_columns_df(df, func_name, params, ser_params):
     for key, value in ser_params.items():
         ser_par.append('{}={}'.format(key, value))
 
-    s_par = '{}'.format(', '.join(ser_par[:]))
+    s_par = '{}'.format(', '.join(ser_par))
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: Why do you need to format the string? Why not do so:
`s_par = ', '.join(ser_par)`

File: sdc/datatypes/hpat_pandas_dataframe_functions.py (hunk 3 above)
Comment (densmirn) [reply sdc/datatypes/hpat_pandas_dataframe_functions.py]: > Looks like `numpy` isn't used at all. So please remove that.

Please apply the comment.

File: sdc/tests/test_dataframe.py [hunk 9]
```diff
@@ -1357,6 +1357,23 @@ def test_impl(df):
...
+
+        hpat_func = sdc.jit(test_impl)
+        df = pd.DataFrame({"A": [.2, .0, .6, .2],
+                           "B": [.5, .6, .7, .8],
+                           "C": [2, 0, 6, 2],
+                           "D": [.2, .1, np.nan, .5],
+                           "E": [-1, np.nan, 1, np.inf],
+                           "F": [np.nan, np.nan, np.inf, np.nan]})
+        for periods in [0, 1, 2, 5, 10, -1, -2, -5]:
+            for method in [None, 'pad', 'ffill', 'backfill', 'bfill']:
```
Comment (densmirn) [inline sdc/tests/test_dataframe.py]: I propose to use `unittest.subTest` for such cases. Moreover if you will use `itertools.product` you can collapse 2 loops to a single one.

File: sdc/tests/test_dataframe.py [hunk 10]
```diff
@@ -1369,10 +1370,12 @@ def test_impl(df, periods, method):
                            "E": [-1, np.nan, 1, np.inf],
                            "F": [np.nan, np.nan, np.inf, np.nan]})
         for periods in [0, 1, 2, 5, 10, -1, -2, -5]:
-            for method in [None, 'pad', 'ffill', 'backfill', 'bfill']:
-                result_ref = test_impl(df, periods, method)
-                result = hpat_func(df, periods, method)
-                pd.testing.assert_frame_equal(result, result_ref)
+            with self.subTest(periods=periods):
+                for method in [None, 'pad', 'ffill', 'backfill', 'bfill']:
+                    with self.subTest(method=method):
```
Comment (densmirn) [inline sdc/tests/test_dataframe.py]: Let's merge 2 usages of `subTest` to a single one:
`with self.subTest(periods=periods, method=method):`

and let's use `itertools.product`

File: sdc/tests/test_dataframe.py [hunk 11]
```diff
@@ -1369,13 +1370,12 @@ def test_impl(df, periods, method):
...
                            "F": [np.nan, np.nan, np.inf, np.nan]})
-        for periods in [0, 1, 2, 5, 10, -1, -2, -5]:
-            with self.subTest(periods=periods):
-                for method in [None, 'pad', 'ffill', 'backfill', 'bfill']:
-                    with self.subTest(method=method):
-                        result_ref = test_impl(df, periods, method)
-                        result = hpat_func(df, periods, method)
-                        pd.testing.assert_frame_equal(result, result_ref)
+        params = list(product(*[[0, 1, 2, 5, 10, -1, -2, -5], [None, 'pad', 'ffill', 'backfill', 'bfill']]))
+        for periods, method in params:
```
Comment (densmirn) [inline sdc/tests/test_dataframe.py]:
```suggestion
all_periods = [0, 1, 2, 5, 10, -1, -2, -5]
methods = [None, 'pad', 'ffill', 'backfill', 'bfill']
for periods, method in product(all_periods, methods):
```

**Comment by AlexanderKalistratov** [comment]
@Rubtsowa merge conflict

**Pull request merged**